tests/test_permutation_strategies.py

```diff
@@ -0,0 +1,199 @@
+# BSD 2-Clause License
+#
+# Copyright (c) 2021-2024, Hewlett Packard Enterprise
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import dataclasses
+
+import pytest
+
+from smartsim.entity import strategies
+from smartsim.entity.strategies import ParamSet
+from smartsim.error import errors
+
+pytestmark = pytest.mark.group_a
+
+
+def test_strategy_registration(monkeypatch):
+    monkeypatch.setattr(strategies, "_REGISTERED_STRATEGIES", {})
+    assert not strategies._REGISTERED_STRATEGIES
+
+    new_strat = lambda params, exe_args, nmax: []
+    strategies._register("new_strat")(new_strat)
+    assert strategies._REGISTERED_STRATEGIES == {"new_strat": new_strat}
+
```

Comment:
Just to check my understanding! 

Here you first make sure that _REGISTERED_STRAT is empty. next you define a lambda function (easier than actually writing a user function) then you register it. But it seems kinda funny to do `._register("new_strat")(new_strat)` - why the two ()() when _register just accpets (str)

Reply:
^^ Exactly correct!!

The reason for the two `()` calls is because `register` is a callable that returns a callable that actually does the registration. The registration is actually done during the second call. You can think of it as actually doing something like this

```py
_tmp_fn_to_add_a_new_strategy_under_the_name_new_strategy = strategies._register("new_strat")

reveal_type(_tmp_fn_to_add_a_new_strategy_under_the_name_new_strategy)
# reveals: `t.Callable[[PermutationStrategyType], PermutationStrategyType]`

assert strategies._REGISTERED_STRATEGIES == {}
# Passes! Notice that nothing has been added yet

new_strat = lambda p, e, n: []
return_from_tmp_fn = _tmp_fn_to_add_a_new_strategy_under_the_name_new_strategy(new_strat)

assert return_from_tmp_fn is new_strat
# Passes! Notice that the function is returned unchanged

assert strategies._REGISTERED_STRATEGIES == {"new_strat": new_strat}
# Passes! Notice that the key was only added after
# `_tmp_fn_to_add_a_new_strategy_under_the_name_new_strategy` was called
```

All of this is for the weird magic that let's us use this function like a decorator. This test is functionally testing this use case
```py
stategies._register("new_strat")
def new_strat(p, e, n):
    return []
```